### PR TITLE
indent option on breakable

### PR DESCRIPTION
**Pull request by kddnewton**
This commit adds the ability to toggle whether or not a Breakable will
indent when it is printed. In order to achieve this, we delay pushing
anything into the output buffer until flush is called. When it is
called, we iterate through the doc nodes and check at each step if they
fit into the current line (as we did before). Now, however, when a
Breakable is printed, it checks if it should indent itself or not.

These changes open the door for many more changes that can be
accomplished now that printing is delayed until flush is called. Because
of the delay:

* nodes in the doc tree can now change their representation depending on
  whether or not the current group is broken
* we can force all parent groups to break up the tree, instead of just
  the current group. This will allow all of the nodes up the tree to
  break and change their representation as well
* by passing the indent level through the tree as its being printed
  instead of keeping a single variable around, we can more tightly
  control indentation level as the tree is printed.

### CODE REVIEW

File: lib/prettyprint.rb [hunk 1]
```diff
@@ -42,7 +41,7 @@ class PrettyPrint
   #     output
   #   end
   #
-  def PrettyPrint.format(output=''.dup, maxwidth=79, newline="\n", genspace=lambda {|n| ' ' * n})
+  def PrettyPrint.format(output=''.dup, maxwidth=80, newline="\n", genspace=lambda {|n| ' ' * n})
```
Comment (tenderlove) [inline lib/prettyprint.rb:44]: Why change this?  Changing to 80 might be fine, but it doesn't seem related to the purpose of this PR (unless I'm missing something)

File: lib/prettyprint.rb [hunk 2]
```diff
@@ -208,7 +180,6 @@ def text(obj, width=obj.length)
   #
   # If +width+ is not specified, +sep.length+ is used. You will have to
   # specify this when +sep+ is a multibyte character, for example.
-  #
```
Comment (tenderlove) [inline lib/prettyprint.rb]: Also unrelated 😅

File: lib/prettyprint.rb [hunk 3]
```diff
@@ -248,234 +215,289 @@ def breakable(sep=' ', width=sep.length)
...
+    commands = [[IndentLevel.new(genspace: genspace), MODE_BREAK, doc]]
+
+    # This is a small optimization boolean. It keeps track of whether or not
+    # when we hit a group node we should check if it fits on the same line.
+    should_remeasure = false
+
+    # This is a linear stack instead of a mutually recursive call defined on
+    # the individual doc nodes for efficiency.
+    while commands.any?
+      indent, mode, doc = commands.pop
```
Comment (tenderlove) [inline lib/prettyprint.rb]: I think you can pull this up in to the while test like:

```ruby
while (indent, mode, doc = commands.pop)
```

Then you don't need the `any?` call.  (Just a suggestion though, the current code seems fine)

File: lib/prettyprint.rb [hunk 4]
```diff
@@ -248,234 +215,289 @@ def breakable(sep=' ', width=sep.length)
...
-      output_width + @width
+    def initialize
+      @objs = []
+      @width = 0
     end
 
-    # Include +obj+ in the objects to be pretty printed, and increment
-    # this Text object's total width by +width+
-    def add(obj, width)
+    def add(obj: "", width: obj.length)
```
Comment (tenderlove) [inline lib/prettyprint.rb]: Does this method signature need to change?  Also why remove the comment?  The comment seems accurate afaict.

**Review by tenderlove — COMMENTED** [review]
I left some comments.  I don't really understand how these objects work together, so it's hard for me to make a deep technical analysis, so all of my comments are pretty "surface level".

If this PR passes tests, then it's probably fine IMO. 😅

File: lib/prettyprint.rb (hunk 1 above)
Comment (kddnewton) [reply lib/prettyprint.rb:44]: It was an off-by-one thing. Previously it was something like `< maxwidth` and now it boils down to `<= maxwidth` but it's a little different because we're subtracting remaining space. It just made it a little nicer to express this way.

File: lib/prettyprint.rb (hunk 2 above)
Comment (kddnewton) [reply lib/prettyprint.rb]: Ahh shoot I'll fix that.

File: lib/prettyprint.rb (hunk 3 above)
Comment (kddnewton) [reply lib/prettyprint.rb]: Nice yes I will do that

File: lib/prettyprint.rb (hunk 4 above)
Comment (kddnewton) [reply lib/prettyprint.rb]: Ahh shoot no it doesn't need to change, I'll revert that.

**Comment by kddnewton** [comment]
Thanks @tenderlove I've updated from your feedback.

File: lib/prettyprint.rb (hunk 1 above)
Comment (tenderlove) [reply lib/prettyprint.rb:44]: Ahhh, interesting.  Thanks for the context!
